Fix cluster queue validating webhook

kannon92 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind bug
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

When there is a mismatch between coveredResources and Flavors you can hit a panic.  
#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #786 

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
Fix panic in cluster queue if resources and coveredResources do not have the same length.
```

kannon92 commented:
/cancel hold

kannon92 commented:
/hold cancel

tenzen-y commented on apis/kueue/webhooks/clusterqueue_webhook.go:
```diff
@@ -156,6 +156,7 @@ func validateFlavorQuotas(flavorQuotas kueue.FlavorQuotas, coveredResources []co
 	allErrs := validateNameReference(string(flavorQuotas.Name), path.Child("name"))
 	if len(flavorQuotas.Resources) != len(coveredResources) {
 		allErrs = append(allErrs, field.Invalid(path.Child("resources"), field.OmitValueType{}, "must have the same number of resources as the coveredResources"))
+		return allErrs
 	}
 
 	for i, rq := range flavorQuotas.Resources {
```
```suggestion
	}

	coveredResNum := len(coveredResources)
	for i, rq := range flavorQuotas.Resources {
		if coveredResNum < i+1 {
			break
		}
```

Validating matched resources would be helpful.

kannon92 replied:
Done.

tenzen-y commented:
Also, I think this is a user-facing bug. So can you update a release note?

kannon92 commented:
> Also, I think this is a user-facing bug. So can you update a release note?

I believe I have addressed all your comments.  This is ready for another review.

tenzen-y reviewed (commented):
@kannon92 Thank you for the updates!

/lgtm
/assign @alculquicondor

alculquicondor commented on apis/kueue/webhooks/clusterqueue_webhook.go:
```diff
@@ -158,7 +158,11 @@ func validateFlavorQuotas(flavorQuotas kueue.FlavorQuotas, coveredResources []co
 
+	coveredResNum := len(coveredResources)
 	for i, rq := range flavorQuotas.Resources {
+		if coveredResNum < i+1 {
```
nit (just my personal taste)

```suggestion
		if i >= len(coveredResNum) {
```

I wouldn't imagine that `len` is costly. slices are basically a struct of one pointer and two numbers (len and cap).

alculquicondor reviewed (commented):
/lgtm
/approve
/hold
(for nit)

alculquicondor commented on apis/kueue/webhooks/clusterqueue_webhook.go:
```diff
@@ -158,7 +158,11 @@ func validateFlavorQuotas(flavorQuotas kueue.FlavorQuotas, coveredResources []co
 		allErrs = append(allErrs, field.Invalid(path.Child("resources"), field.OmitValueType{}, "must have the same number of resources as the coveredResources"))
 	}
 
+	coveredResNum := len(coveredResources)
```
```suggestion
```

kannon92 replied:
Thats what I get for being lazy..  I pushed up a fix for this.

alculquicondor commented:
/lgtm
/approve
/label tide/merge-method-squash

kannon92 commented:
/hold cancel

alculquicondor commented:
/cherry-pick release-0.3

alculquicondor commented:
@kannon92 could you open a manual cherry-pick?

kannon92 commented:
> @kannon92 could you open a manual cherry-pick?

Sure.  Is this something that explains what I need to do for this?

alculquicondor commented:
No, I just tried to do automatic cherry-pick on `release-0.3` and it had a conflict.

alculquicondor commented:
Not sure if using this script might help https://github.com/kubernetes-sigs/kueue/blob/main/hack/cherry_pick_pull.sh

But doing it completely manually should be fine.

alculquicondor commented:
/milestone v0.4
